### ISSUE TITLE
JavaScript Ice/ami assert

### ISSUE THREAD

**Issue by pepone** [zeroc-ice/ice#3087]
```
testing AsyncResult operations... Trace: Assertion failed: undefined
    at Debug.assert (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/Debug.js:8:21)
    at OutgoingAsync.markSent (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/AsyncResult.js:47:15)
    at OutgoingAsync.markSent (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/OutgoingAsync.js:149:24)
    at OutgoingAsync.sent (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/OutgoingAsync.js:377:14)
    at OutgoingMessage.sent (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/ConnectionI.js:1504:27)
    at ConnectionI.sendNextMessage (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/ConnectionI.js:1156:25)
    at ConnectionI.message (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/ConnectionI.js:566:26)
    at TcpTransceiver._bytesWrittenCallback (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/ConnectionI.js:160:28)
    at file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/TcpTransceiver.js:155:30
    at afterWrite (node:internal/streams/writable:710:5)
```

Got this on a local machine while investigating #2979 , likely related to the same underlying issue.

**Comment by pepone** [comment]
Added a message to the assert, and it shows the request was already completed when sent was called:

```
Trace: Assertion failed: Unexpected async result status: 3 completed: Error
    at OutgoingAsync.markFinished (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/AsyncResult.js:63:35)
    at OutgoingAsync.completed (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/OutgoingAsync.js:499:18)
    at ConnectionI.upcall (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/ConnectionI.js:596:31)
    at ConnectionI.message (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/ConnectionI.js:574:14)
    at Immediate._onImmediate (file:///C:/Users/jose/source/repos/3.8/ice/js/node_modules/ice/src/Ice/ConnectionI.js:577:43)
    at process.processImmediate (node:internal/timers:491:21), [object Promise]
```

The `status: 3` correspond with `AsyncResul.Ok | AsyncResult.Done`

And the stack shows the completion was done from the receiving of the reply upcall.

**Comment by pepone** [comment]
I think this is similar to the retry issue recently fixed in C# #2722.

If the request is sent synchronously we mark the request as sent right away which seems correct:

https://github.com/zeroc-ice/ice/blob/e56be0257dc4e0a6f5e426f7c1e9ab442e4b0587/js/src/Ice/ConnectionI.js#L1230-L1234

When the request is sent synchronously we only mark it as sent from `sendNextMessage`

https://github.com/zeroc-ice/ice/blob/e56be0257dc4e0a6f5e426f7c1e9ab442e4b0587/js/src/Ice/ConnectionI.js#L1154-L1156

But this is too late as we have already sent the whole buffer

https://github.com/zeroc-ice/ice/blob/e56be0257dc4e0a6f5e426f7c1e9ab442e4b0587/js/src/Ice/TcpTransceiver.js#L152-L157
